### DISCUSSION TITLE
How to generate migrations when you defined models and config and migration in custom path.

### DISCUSSION THREAD

**Discussion by tirathsharma098** [sequelize/sequelize#15445] [Help & Questions]
How to generate migrations when you defined models and config and migration in custom path.
---------------------------------------
Everything is working perfect. 
Migration is generating but without the models details that I have defined.
#### .sequelizerc
```
const path = require("path");
module.exports = {
    "config": path.resolve("src/database", "config.json"),
    "models-path": path.resolve("src", "models"),
    "seeders-path": path.resolve("src", "seeds"),
    "migrations-path": path.resolve("src", "migrations"),
    debug: true,
};
```
#### config.json  note: I am using my db credential correctly
```
{
  "development": {
    "username": "root",
    "password": "passowrd",
    "database": "monitor",
    "host": "127.0.0.1",
    "dialect": "postgres"
  },
  "test": {
    "username": "root",
    "password": "passowrd",
    "database": "monitor",
    "host": "127.0.0.1",
    "dialect": "postgres"
  },
  "production": {
    "username": "root",
    "password": "passowrd",
    "database": "monitor",
    "host": "127.0.0.1",
    "dialect": "postgres"
  }
}
```
#### .db
```
import dotenv from "dotenv";
dotenv.config();
import { Sequelize } from "sequelize";
const dbHost = process.env.DB_HOST;
const dbPort = process.env.DB_PORT;
const dbUserName = process.env.DB_USERNAME;
const dbPassword = process.env.DB_PASSWORD;
const dbName = process.env.DB_NAME;

const sequelize = new Sequelize(dbName, dbUserName, dbPassword, {
    host: dbHost,
    port: Number(dbPort),
    dialect: "postgres",
    logging: (msg) => console.log(msg),
    pool: {
        max: 5,
        min: 0,
        acquire: 30000,
        idle: 10000,
    },
    define: {
        freezeTableName: true,
    },
});

const db = {};
db.Sequelize = Sequelize;
db.sequelize = sequelize;

export default db;
```
#### models -> User.js
```
import { DataTypes } from "sequelize";
import { USER_TYPE } from "../utils/constants.js";
const User = (sequelize) => {
    return sequelize.define(
        "User",
        {
            id: {
                type: DataTypes.UUID,
                allowNull: false,
                primaryKey: true,
                default: DataTypes.UUIDV4,
                field: "id",
            },
            fullName: {
                type: DataTypes.STRING,
                allowNull: false,
                field: "full_name",
            },
            username: {
                type: DataTypes.STRING(65),
                allowNull: false,
                unique: true,
                field: "username",
            },
            password: {
                type: DataTypes.STRING(25),
                allowNull: false,
                field: "password",
            },
            email: {
                type: DataTypes.STRING,
                allowNull: false,
                validate: {
                    isEmail: "Wrong Email Id provided.",
                    notEmpty: "Email should not be empty.",
                },
                field: "email",
            },
            lastLogin: {
                type: DataTypes.DATE,
                defaultValue: DataTypes.NOW,
                allowNull: false,
                field: "last_login",
            },
            tillTokenValidated: {
                type: DataTypes.DATE,
                allowNull: false,
                field: "till_token_validated",
            },
            intro: {
                type: DataTypes.STRING(500),
                allowNull: true,
                defaultValue: "I mean it, and you should also mean it",
                field: "intro",
            },
            score: {
                type: DataTypes.INTEGER,
                defaultValue: 0,
                allowNull: false,
                field: "score",
            },
            userType: {
                type: DataTypes.ENUM(...Object.values(USER_TYPE)),
                defaultValue: USER_TYPE.VIEW,
                allowNull: false,
                field: "user_type",
            },
        },
        {
            tableName: "users",
        }
    );
};

export { User };
```

### project structure.
```
.sequelizerc
index.js
src -> models -> User.js
          api
          database -> db.js, config.json
          seeds
          migrations
          models
```
<hr>
Now when I am running `npx sequelize migration:generate --name user` it is creating migration file which looks like,

```
'use strict';
/** @type {import('sequelize-cli').Migration} */
module.exports = {
  async up (queryInterface, Sequelize) {
    /**
     * Add altering commands here.
     *
     * Example:
     * await queryInterface.createTable('users', { id: Sequelize.INTEGER });
     */
  },

  async down (queryInterface, Sequelize) {
    /**
     * Add reverting commands here.
     *
     * Example:
     * await queryInterface.dropTable('users');
     */
  }
};
```
> So, as you can see this migration does't have my model that I defined.
> I checked so many questions and gone through all detailed description by sequelize in its docs.
> but did't found anything, If you know the solution then please help.
**Sequelize version 6 latest**

**Comment by tirathsharma098** [comment]
> ## How to generate migrations when you defined models and config and migration in custom path.
> Everything is working perfect. Migration is generating but without the models details that I have defined.
> 
> #### .sequelizerc
> ```
> const path = require("path");
> module.exports = {
>     "config": path.resolve("src/database", "config.js"),
>     "models-path": path.resolve("src", "models"),
>     "seeders-path": path.resolve("src", "seeds"),
>     "migrations-path": path.resolve("src", "migrations"),
>     debug: true,
> };
> ```
> 
> #### .config
> ```
> import dotenv from "dotenv";
> dotenv.config();
> import { Sequelize } from "sequelize";
> const dbHost = process.env.DB_HOST;
> const dbPort = process.env.DB_PORT;
> const dbUserName = process.env.DB_USERNAME;
> const dbPassword = process.env.DB_PASSWORD;
> const dbName = process.env.DB_NAME;
> 
> const db = new Sequelize(dbName, dbUserName, dbPassword, {
>     host: dbHost,
>     port: Number(dbPort),
>     dialect: "postgres",
>     logging: (msg) => console.log(msg),
>     pool: {
>         max: 5,
>         min: 0,
>         acquire: 30000,
>         idle: 10000,
>     },
>     define: {
>         freezeTableName: true,
>     },
> });
> 
> export { db };
> ```
> 
> #### models -> User.js
> ```
> import { DataTypes } from "sequelize";
> import { db } from "../database/config.js";
> import { USER_TYPE } from "../utils/constants.js";
> const User = db.define(
>     "User",
>     {
>         id: {
>             type: DataTypes.UUID,
>             allowNull: false,
>             primaryKey: true,
>             default: DataTypes.UUIDV4,
>             field: "id",
>         },
>         fullName: {
>             type: DataTypes.STRING,
>             allowNull: false,
>             field: "full_name",
>         },
>         userType: {
>             type: DataTypes.ENUM(...Object.values(USER_TYPE)),
>             defaultValue: USER_TYPE.VIEW,
>             allowNull: false,
>             field: "user_type",
>         },
>     },
>     {
>         tableName: "users",
>     }
> );
> 
> // export { User };
> export default User;
> ```
> 
> ### project structure.
> ```
> .sequelizerc
> index.js
> src -> models -> User.js
>           api
>           database -> config.js
>           seeds
>           migrations
> ```
> 
> Now when I am running `npx sequelize migration:generate --name user` it is creating migration file which looks like,
> ```
> 'use strict';
> /** @type {import('sequelize-cli').Migration} */
> module.exports = {
>   async up (queryInterface, Sequelize) {
>     /**
>      * Add altering commands here.
>      *
>      * Example:
>      * await queryInterface.createTable('users', { id: Sequelize.INTEGER });
>      */
>   },
> 
>   async down (queryInterface, Sequelize) {
>     /**
>      * Add reverting commands here.
>      *
>      * Example:
>      * await queryInterface.dropTable('users');
>      */
>   }
> };
> ```
> 
> > So, as you can see this migration does't have my model that I defined.
> > I checked so many questions and gone through all detailed description by sequelize in its docs.
> > but did't found anything, If you know the solution then please help.
> > **Sequelize version 6 latest**
Hi @jharting @treythomas123 @xdarklight  would you like to answer please.
Similar to question
https://github.com/sequelize/sequelize/discussions/14627#discussioncomment-4391126

**Comment by tirathsharma098** [comment]
With this update, when using sync force true, tables are creating in db, but after I deleted db and created it again fresh then runing generate migration command is generating empty migration file.

**Reply by paul-uz** [reply]
~~try this https://sequelize.org/docs/v6/other-topics/migrations/#creating-the-first-model-and-migration~~

EDIT - sorry, I see you want to write your model and have the migration generated from it. I'm after the same thing, but can't find anything. The link I provided is to create a model & migration at the same time, but its quite limited.
